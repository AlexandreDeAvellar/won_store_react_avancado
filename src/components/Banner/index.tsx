import Button from '../Button'
import * as S from './styles'
import Ribbon, { RibbonColors, RibbonSizes } from '../Ribbon'
import Image from 'next/image'
import parse from 'html-react-parser'

export type BannerProps = {
  img: string
  title: string
  subtitle: string
  buttonLabel: string
  buttonLink: string
  ribbon?: React.ReactNode
  ribbonColor?: RibbonColors
  ribbonSize?: RibbonSizes
}

const Banner = ({ img, title, subtitle, buttonLink, buttonLabel, ribbon, ribbonColor = 'primary', ribbonSize = 'normal' }: BannerProps) => (
  <S.Wrapper>
    {!!ribbon && (
      <Ribbon color={ribbonColor} size={ribbonSize}>
        {ribbon}
      </Ribbon>
    )}

    <S.ImageWrapper>{<Image src={img} alt={title} fill />}</S.ImageWrapper>

    <S.Caption>
      <S.Titile>{title}</S.Titile>
      <S.Subitile>{parse(subtitle)}</S.Subitile>
      <Button as="a" href={buttonLink} size="large">
        {buttonLabel}
      </Button>
    </S.Caption>
  </S.Wrapper>
)
export default Banner
