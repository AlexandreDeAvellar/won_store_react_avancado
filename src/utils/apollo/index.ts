import { ApolloClient, HttpLink, NormalizedCacheObject } from '@apollo/client'
import { setContext } from '@apollo/client/link/context'
import { useMemo } from 'react'
import { apolloCache } from './apollo-cache'
import { Session } from 'next-auth'

let apolloClient: ApolloClient<NormalizedCacheObject> | null

const createApolloClient = (session?: Session | null) => {
  // https://www.apollographql.com/docs/react/networking/authentication/

  const httpLink = new HttpLink({ uri: `${process.env.NEXT_PUBLIC_API_URL}/graphql` })
  const authLink = setContext((_, { headers, session: clientSession }) => {
    const jwt = session?.jwt || clientSession?.jwt || ''
    const authorization = jwt ? `Bearer ${session?.jwt}` : ''
    return {
      headers: { ...headers, authorization }
    }
  })
  return new ApolloClient({
    link: authLink.concat(httpLink),
    ssrMode: typeof window === 'undefined',
    cache: apolloCache
  })
}

export const initializeApollo = (initialState = null, session?: Session | null) => {
  // serve para verificar se já existe uma instância, para não criar outra
  const apoloClientGlobal = apolloClient ?? createApolloClient(session)

  // se a página usar o apolloClient no lado client
  // hidratamos o estado inicial aqui
  if (initialState) {
    apoloClientGlobal.cache.restore(initialState)
  }

  // sempre inicializando no SSR com cache limpo
  if (typeof window === 'undefined') return apoloClientGlobal

  // cria o apolloClient se estiver no client side
  apolloClient = apolloClient ?? apoloClientGlobal

  return apolloClient
}

export const useApollo = (initialState = null, session?: Session | null) => {
  const store = useMemo(() => initializeApollo(initialState, session), [initialState, session])
  return store
}
