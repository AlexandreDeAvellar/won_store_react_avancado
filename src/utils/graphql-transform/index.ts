export * from './games'
export * from './banner/banner'
export * from './game-by-slug/game-by-slug'
export * from './sections'
export * from './highlight'
export * from './cart'
export * from './orders'
