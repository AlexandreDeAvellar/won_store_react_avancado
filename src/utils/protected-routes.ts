import { GetServerSidePropsContext } from 'next'
import { getServerSession } from 'next-auth'
import { getSession } from 'next-auth/react'
import { nextAuthOptions } from '../pages/api/auth/[...nextauth]'

export const protectedRoutes = async (context: GetServerSidePropsContext) => {
  const session = await getSession(context)

  if (!session) {
    context.res.writeHead(302, { Location: `/sign-in?callbackUrl=${context.resolvedUrl}` })
    context.res.end()
  }

  return session
}

export const protectedRoutesServer = async ({ req, res, resolvedUrl }: GetServerSidePropsContext) => {
  const session = await getServerSession(req, res, nextAuthOptions)

  if (!session) {
    res.writeHead(302, { Location: `/sign-in?callbackUrl=${resolvedUrl}` })
    res.end()
  }

  return session
}
