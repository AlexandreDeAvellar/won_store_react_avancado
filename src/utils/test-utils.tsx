import { ReactNode } from 'react'

import { CartContext, CartContextProps, defaultCartItem } from '../hooks/use-cart'
import { RenderOptions, render } from '@testing-library/react'

import { ThemeProvider } from 'styled-components'
import theme from '../styles/theme'

export type CustomRenderProps = {
  cartProviderProps?: CartContextProps
} & Omit<RenderOptions, 'queries'>

export const customRender = (ui: ReactNode, { cartProviderProps = defaultCartItem, ...renderOptions }: CustomRenderProps = {}) =>
  render(
    <ThemeProvider theme={theme}>
      <CartContext.Provider value={cartProviderProps}>{ui}</CartContext.Provider>
    </ThemeProvider>,
    renderOptions
  )

export * from '@testing-library/react'
export { customRender as render }
