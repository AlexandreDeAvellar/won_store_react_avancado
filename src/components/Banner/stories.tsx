import Banner, { BannerProps } from '.'
import { bannerProps } from './banner-mocks'
import { Story, Meta } from '@storybook/react'

export default {
  title: 'Banner',
  component: Banner,
  argTypes: { ribbon: { type: 'string' } },
  args: bannerProps,
  parameters: { layout: 'fullscreen' }
} as Meta

export const Basic: Story<BannerProps> = (args) => <Banner {...args} />

export const WithRibbon: Story<BannerProps> = (args) => (
  <div style={{ maxWidth: '80rem', maxHeight: '20rem', position: 'relative', backgroundColor: '#ff0000' }}>
    <Banner {...args} />
  </div>
)

WithRibbon.args = {
  ribbon: '25% off',
  ribbonSize: 'normal',
  ribbonColor: 'primary'
}
