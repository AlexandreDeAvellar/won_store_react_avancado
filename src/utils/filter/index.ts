import { ItemProps } from 'components/ExploreSidebar'
import { ParsedUrlQueryInput } from 'querystring'

type ParseArgs = {
  queryString: ParsedUrlQueryInput
  filterItems: Pick<ItemProps, 'type' | 'name'>[]
}

export const parseQueryStringToWhere = ({ queryString, filterItems }: ParseArgs) => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const obj: any = {}

  Object.keys(queryString)
    .filter((item) => item !== 'sort')
    .forEach((key) => {
      const item = filterItems?.find((item) => item.name === key)
      // const type = item?.type === 'checkbox'

      const type = item?.type === 'checkbox' ? 'checkbox' : 'radio'
      const format = {
        checkbox: { name: { in: queryString[key] } },
        radio: { lte: Number(queryString[key]) }
      }
      obj[key.replace('_lte', '')] = format[type]
    })

  return obj
}

export const parseQueryStringToFilter = ({ queryString, filterItems }: ParseArgs) => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const obj: any = {}

  Object.keys(queryString).forEach((key) => {
    const item = filterItems?.find((item) => item.name === key)
    const isCheckbox = item?.type === 'checkbox'
    const isArray = Array.isArray(queryString[key])

    obj[key] = !isArray && isCheckbox ? [queryString[key]] : queryString[key]
  })

  return obj
}
