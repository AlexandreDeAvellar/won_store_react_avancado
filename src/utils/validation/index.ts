import { FormSignUpValues } from '../../components/FormSignUp'
import { ForgotPasswordProps } from '../../components/FormForgotPassword'
import { ResetPasswordProps } from '../../components/FormResetPassword'

import Joi from 'joi'

const fieldsValidations = {
  username: Joi.string().min(5).required(),
  email: Joi.string()
    .email({ tlds: { allow: false } })
    .required(),
  password: Joi.string().required(),
  confirm_password: Joi.string().valid(Joi.ref('password')).required().messages({
    'any.only': 'confirm password does not match with password'
  })
}

export type FieldErros = {
  [key: string]: string
}

function getFieldErrors(objError: Joi.ValidationResult) {
  const errors: FieldErros = {}

  if (objError.error) {
    objError.error.details.forEach((err) => {
      errors[err.path.join('.')] = err.message
    })
  }

  return errors
}

export function signUpValidate(values: FormSignUpValues) {
  const schema = Joi.object(fieldsValidations)
  const result = getFieldErrors(schema.validate(values, { abortEarly: false }))
  return result
}

export function signInValidate(values: Omit<FormSignUpValues, 'username'>) {
  const { email, password } = fieldsValidations
  const schema = Joi.object({ email, password })

  const result = getFieldErrors(schema.validate(values, { abortEarly: false }))
  return result
}

export function forgotValidate(values: ForgotPasswordProps) {
  const { email } = fieldsValidations
  const schema = Joi.object({ email })
  const result = getFieldErrors(schema.validate(values, { abortEarly: false }))
  return result
}

export function resetValidate(values: ResetPasswordProps) {
  const { password, confirm_password } = fieldsValidations
  const schema = Joi.object({ password, confirm_password })
  const result = getFieldErrors(schema.validate(values, { abortEarly: false }))
  return result
}
