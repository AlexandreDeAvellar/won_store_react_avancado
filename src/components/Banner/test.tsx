import Banner, { BannerProps } from '.'
import { bannerProps } from './banner-mocks'
import { render as renderWithTheme, screen } from '../../utils/test-utils'

const props: BannerProps = bannerProps

describe('<Banner />', () => {
  it('should render the image', () => {
    const { container } = renderWithTheme(<Banner {...props} />)

    const img = screen.getByRole('img')
    expect(img.getAttribute('src')).toBe(props.img)

    expect(container).toMatchSnapshot()
  })

  it('should render components correctly', () => {
    renderWithTheme(<Banner {...props} />)

    expect(screen.getByText(/Defy death/i)).toBeInTheDocument()
    expect(screen.getByText(/Play the new /i)).toBeInTheDocument()
  })

  it('should render a Ribbon', () => {
    renderWithTheme(<Banner {...props} ribbon="20% off" ribbonSize="small" ribbonColor="secondary" />)
    expect(screen.getByText(/20% off/i)).toBeInTheDocument()
    expect(screen.getByText(/20% off/i)).toHaveStyle({ backgroundColor: '#3CD3C1' })
    expect(screen.getByText(/20% off/i)).toHaveStyle({ right: '-1.3rem' })
  })
})
