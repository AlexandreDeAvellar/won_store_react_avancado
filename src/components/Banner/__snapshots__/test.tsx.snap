// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Banner /> should render the image 1`] = `
.c5 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background: linear-gradient(180deg,#ff5f5f 0%,#f062c0 50%);
  color: #FAFAFA;
  border: 0;
  border-radius: 0.4rem;
  padding: 0.4rem;
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
  height: 3.7rem;
  font-size: 1rem;
  padding: 0.4rem 2.8rem;
}

.c5:focus {
  outline: 1px dashed;
}

.c5:hover {
  background: linear-gradient(180deg,#e35565 0%,#d958a6 50%);
  background: linear-gradient(180deg,#e35565 0%,#d958a6 50%);
}

.c0 {
  position: relative;
}

.c1 {
  width: 100%;
  height: 23rem;
  background-color: #EAEAEA;
  background-position: center center;
  position: relative;
}

.c2 {
  width: 100%;
  background-color: rgba(0,0,0,0.7);
  padding: 1.2rem;
}

.c3 {
  font-size: 1.2rem;
  font-weight: 600;
  color: #FAFAFA;
}

.c4 {
  font-size: 0.8rem;
  font-weight: 400;
  color: #FAFAFA;
  margin-bottom: 0.8rem;
}

.c4 strong {
  color: #F231A5;
  font-weight: 600;
}

@media (min-width:768px) {
  .c0 {
    box-shadow: 0 0.4rem 0.5rem 0 rgba(0,0,0,0.2);
  }
}

@media (max-width:1170px) {

}

@media (min-width:768px) {
  .c1 {
    height: 58rem;
  }
}

@media (min-width:768px) {
  .c2 {
    border-radius: 0 0 0.4rem 0.4rem;
    padding: 2.0rem;
    position: absolute;
    bottom: 0;
    left: 0;
  }
}

@media (min-width:768px) {
  .c3 {
    font-size: 2rem;
  }
}

@media (min-width:768px) {
  .c4 {
    font-size: 1.2rem;
  }
}

<div>
  <main
    class="c0"
  >
    <div
      class="c1"
    >
      <img
        alt="Defy death"
        data-nimg="fill"
        decoding="async"
        loading="lazy"
        sizes="100vw"
        src="/_next/image?url=https%3A%2F%2Fsource.unsplash.com%2Fuser%2Fwillianjusten%2F1042x580&w=3840&q=75"
        srcset="/_next/image?url=https%3A%2F%2Fsource.unsplash.com%2Fuser%2Fwillianjusten%2F1042x580&w=640&q=75 640w, /_next/image?url=https%3A%2F%2Fsource.unsplash.com%2Fuser%2Fwillianjusten%2F1042x580&w=750&q=75 750w, /_next/image?url=https%3A%2F%2Fsource.unsplash.com%2Fuser%2Fwillianjusten%2F1042x580&w=828&q=75 828w, /_next/image?url=https%3A%2F%2Fsource.unsplash.com%2Fuser%2Fwillianjusten%2F1042x580&w=1080&q=75 1080w, /_next/image?url=https%3A%2F%2Fsource.unsplash.com%2Fuser%2Fwillianjusten%2F1042x580&w=1200&q=75 1200w, /_next/image?url=https%3A%2F%2Fsource.unsplash.com%2Fuser%2Fwillianjusten%2F1042x580&w=1920&q=75 1920w, /_next/image?url=https%3A%2F%2Fsource.unsplash.com%2Fuser%2Fwillianjusten%2F1042x580&w=2048&q=75 2048w, /_next/image?url=https%3A%2F%2Fsource.unsplash.com%2Fuser%2Fwillianjusten%2F1042x580&w=3840&q=75 3840w"
        style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
      />
    </div>
    <div
      class="c2"
    >
      <h2
        class="c3"
      >
        Defy death
      </h2>
      <h3
        class="c4"
      >
        <p>
          Play the new 
          <strong>
            CrashLands
          </strong>
           season
        </p>
      </h3>
      <a
        class="c5"
        href="/games/defy-death"
      >
        <span>
          Buy now
        </span>
      </a>
    </div>
  </main>
</div>
`;
